packages/workspace/src/browser/workspace-service.ts

```diff
@@ -55,20 +67,172 @@ export class WorkspaceService implements FrontendApplicationContribution {
+        this._workspaceRootChange = {
+            root: undefined,
+            roots: [],
+            workspaceConfigFile: undefined
+        };
+        await this.setupRoots();
+        this.updateTitle();
+        this.deferredWorkspaceRootChange.resolve(this._workspaceRootChange);
+        if (this._workspaceRootChange.workspaceConfigFile) {
+            await this.watchRootConfigFile();
+        } else {
+            const configUri = await this.getWorkspaceConfigFileUri();
+            if (configUri) {
+                this.watcher.onFilesChanged(async event => {
+                    const isRootConfigAdded = !!event.find(change =>
+                        change.type === FileChangeType.ADDED && change.uri.toString() === configUri.toString()
+                    );
+                    const isRootConfigRemoved = !!event.find(change =>
+                        change.type === FileChangeType.DELETED && change.uri.toString() === configUri.toString()
+                    );
+                    if (event) {
+                        if (isRootConfigAdded) {
+                            await this.onRootConfigFileChanged();
+                            await this.watchRootConfigFile();
+                        } else if (isRootConfigRemoved) {
+                            this.stopWatch();
+                            this.startWatch(this._root);
+                            if (this._root) {
+                                this._workspaceRootChange.roots = [this._root];
+                            }
+                            this._workspaceConfigFile = undefined;
+                            this._workspaceRootChange.workspaceConfigFile = undefined;
+                            this.onWorkspaceRootChangeEmitter.fire(this._workspaceRootChange);
+                        }
+                    }
+                });
+            }
+        }
+
+        await this.openWindow({ preserveWindow: true });
+    }
+
+    get root(): Promise<FileStat | undefined> {
+        return this.deferredRoot.promise;
+    }
+    get workspaceRootChange(): Promise<WorkspaceRootChange> {
+        return this.deferredWorkspaceRootChange.promise;
+    }
+    get numberOfRoots(): number {
+        return this._roots.length;
+    }
+
+    protected readonly onWorkspaceRootChangeEmitter = new Emitter<WorkspaceRootChange>();
+    get onWorkspaceRootChanged(): Event<WorkspaceRootChange> {
+        return this.onWorkspaceRootChangeEmitter.event;
+    }
+
+    protected async watchRootConfigFile(): Promise<void> {
+        const configUri = await this.getWorkspaceConfigFileUri();
+        if (configUri) {
+            const configResource = await this.resourceProvider(configUri);
+            if (configResource.onDidChangeContents) {
+                configResource.onDidChangeContents(async () => await this.onRootConfigFileChanged());
+            }
+        }
+    }
+
+    protected async onRootConfigFileChanged(): Promise<void> {
+        this.stopWatch();
```

Comment:
you should find a delta, stop watching removed root and start watching new roots, instead of restarting the whole process